[IMP] website_event, * : empty states

xlu-odoo opened the pull request:
*: website_event_booth, website_event_exhibitor, website_event_track

Before this commit, when no events (or exhibitors, tracks,..) were displayed for the user, only a title (eg. "No events found.") was shown  to them.

Now, it will display an illustration (color adapts to theme's color) alongside more information:
- title 
- call-out 
- call-to-action for editors (add event, tracks, exhibitors..)

task-2489681

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/107638).

xlu-odoo commented:
@stefanorigano & @LNAodoo can you both review accordingly, please?

As discussed with you @LNAodoo , the new buttons on any empty page (editor view) to create new events needs new development from your team: calling a modal instead of a new page in BE.

Thanks 🙏

stefanorigano commented on addons/website_event/views/event_templates_list.xml:
```diff
@@ -149,9 +149,20 @@
-            <div class="alert alert-info text-center" groups="event.group_event_user">
-                <p class="m-0">Use the top button '<b>+ New</b>' to create an event.</p>
+            <div class="d-flex flex-column align-items-center justify-content-center text-center">
+                <div class="mb-3 o_event_svg">
```
`o_event_ ` prefix : In other files you added "website" -> `o_wevent_` .
Both are good but pick one, preferably the one already used by the modules ;)

Classes order: custom classes first, then utility classes.

stefanorigano commented on addons/website_event_exhibitor/views/event_exhibitor_templates_list.xml:
```diff
@@ -121,13 +121,24 @@
-            <div t-if="search_key" class="alert alert-info text-center">
-                <p class="m-0">We did not find any exhibitor matching your <strong t-out="search_key"/> search.</p>
+            <div class="d-flex flex-column align-items-center justify-content-center text-center">
+                <div class="m-3 o_wevent_exhibitor_svg">
```
classes order

stefanorigano commented on addons/website_event/views/event_templates_list.xml:
```diff
@@ -149,9 +149,20 @@
+            </div> 
+            <div class="o_not_editable d-flex flex-column align-items-center justify-content-between my-3" groups="event.group_event_user">
+                <div class="d-flex flex-column align-items-center justify-content-between">
+                    <a class="btn btn-secondary"> <!-- Add link to event modal creation -->
```
`onClick="() => this.onClickNewContent(event)"` ?
Not sure, eventually check with the devs

stefanorigano replied:
You have check this as "resolved", it's a mistake?

stefanorigano reviewed (changes requested):
Hello @xlu-odoo , minor remarks in the code.
Visually ok to me. As said it doesn't match all industries, but it's fine I guess.

xlu-odoo commented:
@stefanorigano

stefanorigano commented on addons/website_event/views/event_svg.xml:
```diff
@@ -0,0 +1,56 @@
+<odoo>
+    <template id="website_event.event_svg">
+        <svg width="211" height="179" viewBox="0 0 211 179" fill="none" xmlns="http://www.w3.org/2000/svg">
+            <g id="Frame" clip-path="url(#clip0_56_106)">
```
I don't think we need `id="Frame"`.
Could you also rename `clip0_56_106` to something less misterius ? ^^

Not sure if is git, but the indentation here is wrong

stefanorigano commented on addons/website_event/views/event_svg.xml:
```diff
@@ -0,0 +1,56 @@
+    <template id="website_event.event_svg">
+        <svg width="211" height="179" viewBox="0 0 211 179" fill="none" xmlns="http://www.w3.org/2000/svg">
+            <g id="Frame" clip-path="url(#clip0_56_106)">
+            <g fill="var(--o-color-1)" opacity="1">
```
Provide a fallback for all these `fill` attributes so we can debug outside of the website environment, eg:

```suggestion
            <g fill="var(--o-color-1, #000)" opacity="1">
```

stefanorigano commented on addons/website_event/views/event_svg.xml:
```diff
@@ -0,0 +1,56 @@
+            <path d="M75.69 93.29C75.33 93.29 74.98 93.14 74.72 92.85L73.04 90.92C72.57 90.39 72.63 89.57 73.17 89.11C73.7 88.65 74.51 88.71 74.98 89.24L76.66 91.18C77.13 91.71 77.07 92.53 76.53 92.99C76.29 93.2 75.99 93.3 75.69 93.3V93.29Z" />
+            <path d="M116.05 139.79C115.69 139.79 115.34 139.64 115.08 139.35L113.4 137.42C112.94 136.89 113 136.07 113.53 135.61C114.06 135.14 114.88 135.2 115.34 135.74L117.02 137.67C117.49 138.2 117.43 139.02 116.89 139.48C116.65 139.69 116.35 139.79 116.05 139.79Z" />
+            </g>
+            <path opacity="1" d="M137.79 54.44C133.73 59.54 126.31 60.38 121.21 56.33C116.11 52.27 115.27 44.85 119.32 39.75V39.73C119.59 39.41 119.53 38.94 119.21 38.68L110.14 31.47C106.2 28.34 100.46 28.99 97.33 32.93L26.31 122.23C23.18 126.17 23.83 131.91 27.77 135.04L37.07 142.44C37.45 142.74 37.99 142.67 38.3 142.29C42.36 137.19 49.78 136.35 54.88 140.4C59.98 144.46 60.82 151.88 56.77 156.98 ...
```
We probably don't need all these `opacity="1"`

stefanorigano commented on addons/website_event/views/event_svg.xml:
```diff
@@ -0,0 +1,56 @@
+            </g>
+            <defs>
+            <clipPath id="clip0_56_106">
+            <rect width="210.97" height="179" fill="white"/>
```
Question: will this white rect cover any page's background color?
If yes we should avoid that I guess.

xlu-odoo replied:
Removed it, didn't saw any changes with or without 🤔

stefanorigano commented on addons/website_event_track/views/event_track_templates_agenda.xml:
```diff
@@ -6,14 +6,27 @@
-                </a>
+            <div class="col-12">
+                <div class="d-flex flex-column align-items-center justify-content-center text-center">
+                    <div class="m-3 o_wevent_exhibitor_svg">
```
Custom classes before utility ones
```suggestion
                    <div class="o_wevent_exhibitor_svg m-3">
```

stefanorigano commented on addons/website_event_track/views/event_track_templates_agenda.xml:
```diff
@@ -6,14 +6,27 @@
+                            We couldn't find any track matching your <strong t-out="search_key"/> search.
+                        </t>
+                        <t t-else="">
+                            We couldn't find any track at this momemt.
```
```suggestion
                            We couldn't find any track at this moment.
```

stefanorigano commented on addons/website_event_track/views/event_track_templates_agenda.xml:
```diff
@@ -6,14 +6,27 @@
+                        <t t-call="website_event.event_svg"/>
+                    </div>
+                    <h2 class="mb-2">No track found.</h2>
+                    <p>
```
You could save a bunch of lines and one indentation level applying the condition in the `<p>` tag directly (in the other template too).
```xml

<p t-if="search_key">We couldn't find any track matching your <strong t-out="search_key"/> search.</p>
<p t-else="">We couldn't find any track at this moment.</p>

```

stefanorigano commented on addons/website_event_track/views/event_track_templates_list.xml:
```diff
@@ -117,17 +117,29 @@
             </div>
         </div>
     </t>
+    
```
ok

stefanorigano reviewed (changes requested):
Thanks @xlu-odoo , some remarks for you.

About the commit message, when using `*`  in the commit title you're supposed to list the affected modules in the commit description, eg https://github.com/odoo/enterprise/commit/70adf1ec0025f5c966e8619a84e902ebbad1633a

thiryr commented on addons/website_event/views/event_templates_list.xml:
```diff
@@ -149,9 +149,20 @@
+                    <t t-call="website_event.event_svg"/>
+                </div>
+                <h2 class="mb-2">No events scheduled yet</h2>
+                <p>We couldn’t find any events scheduled at this moment.</p>
```
Can we maybe do a t-if/t-else to say "we couldn't find any event matching your search criteria"? 

This way it's more accurate, and we're consistent with the other messages. We should have access to ``searches['search']`` here.

xlu-odoo replied:
Updated, can you check if it was done the right way?

thiryr reviewed (commented):
Hey there, just a first pass review while I try to think of something nice for the action :)

(woops, seems like stefano already pointed out most of these)

thiryr commented on addons/website_event/views/event_templates_list.xml:
```diff
@@ -149,9 +149,20 @@
+            </div> 
+            <div class="o_not_editable d-flex flex-column align-items-center justify-content-between my-3" groups="event.group_event_user">
+                <div class="d-flex flex-column align-items-center justify-content-between">
+                    <a class="btn btn-secondary"> <!-- Add link to event modal creation -->
```
Update:

After trying a bunch of stuff, and confirming with team website. We cannot open an action from inside the website (the service just isn't available from there).

So you can just use the same thing you used for the others, and just open a new event in the back-end via an action link :)

xlu-odoo replied:
Ok thanks for the research on this.
i'm gonna update it with an action link. 👍

thiryr commented on addons/website_event/views/event_templates_list.xml:
```diff
@@ -147,11 +147,23 @@
+                    <t t-call="website_event.event_svg"/>
+                </div>
+                <h3 class="mb-2">No events scheduled yet</h3>
+                <p t-if="search or searches and searches['search']">We couldn't find any event matching your <strong t-out="search or searches and searches['search']"/> search.</p>
```
(Actually... I'm not sure, I'll check again sorry)

thiryr replied:
Ok, yes. I believe my suggestion is accurate. I'm not sure why we use search above, it's probably a mistake.

Reasoning for not using ``original_search``:
If we didn't find anything, fuzzy_search will be falsy. So original_search will be False, and nothing will appear.

xlu-odoo commented on addons/website_event/views/event_templates_list.xml:
```diff
@@ -147,11 +147,23 @@
+            </div> 
+            <div class="o_not_editable d-flex flex-column align-items-center justify-content-between my-3" groups="event.group_event_user">
+                <div class="d-flex flex-column align-items-center justify-content-between">
+                    <a class="o_wevent_cta btn"> <!-- Add working link to event modal creation -->
```
With your suggestion, I'm going straight to the App drawer/switcher instead of the new event form view

thiryr replied:
Hm, I could have sworn it worked. I'll try to figure out how to create a link for that then. I don't think you can just create a link to the name of the view (because it doesn't define an action).

Ping me if you don't hear anything by the end of the week :)

thiryr replied:
I couldn't find a nice href solution. So I think we'll need to pass a new value to the render, no biggie.

Here's a commit that works:
a594eaa2e2b0f2fbd2b1c4c7d00c8322afffa2f5

thiryr commented on addons/website_event/views/event_templates_list.xml:
```diff
@@ -147,11 +147,23 @@
+            </div> 
+            <div class="o_not_editable d-flex flex-column align-items-center justify-content-between my-3" groups="event.group_event_user">
+                <div class="d-flex flex-column align-items-center justify-content-between">
+                    <a class="o_wevent_cta btn"> <!-- Add working link to event modal creation -->
```
```suggestion
                    <a class="o_wevent_cta btn" href="/web?#model=event.event&view_type=form&action=event.action_event_view">
```

thiryr reviewed (commented):
Last ping for this, I swear ^^'

thiryr commented on addons/website_event_booth/views/event_booth_templates.xml:
```diff
@@ -22,22 +22,38 @@
+                            <p>This event is not open to exhibitors registration at this time</p>
+                        </div>
+                        <div class="o_not_editable d-flex flex-column align-items-center justify-content-between my-3" groups="event.group_event_user">
+                            <div t-if="request.env.user.has_group('event.group_event_manager')" class="d-flex flex-column align-items-center justify-content-between">
```
Is there any reason we don't use ``groups`` here?

xlu-odoo replied:
I'm not sure to understand here sorry

thiryr replied:
No problem I'm asking whether we could do this, or if there's a reason to use ``has_group``
```suggestion
                            <div groups="event.group_event_manager" class="d-flex flex-column align-items-center justify-content-between">
```

xlu-odoo replied:
Not yet familiar with that. It was already set like that. Should I change with your suggestion?

thiryr replied:
Hm, well if it was already like that I guess that's fine. I'm genuinely not sure if there could have been a reason, so let's leave it like that ^^

thiryr commented on addons/website_event/views/event_templates_list.xml:
```diff
@@ -147,11 +147,23 @@
+                    <t t-call="website_event.event_svg"/>
+                </div>
+                <h3 class="mb-2">No events scheduled yet</h3>
+                <p t-if="searches and searches['search']">We couldn't find any event matching your <strong t-out="searches and searches['search']"/> search.</p>
```
Actually this is always set, no need to check if it exists
```suggestion
                <p t-if="searches['search']">We couldn't find any event matching your <strong t-out="searches['search']"/> search.</p>
```

thiryr replied:
```suggestion
                <p t-if="searches and searches['search']">We could not find any event matching your <strong t-out="searches and searches['search']"/> search.</p>
```
I feel like we generally want to avoid contractions in UI, but I'm no UI expert so that's just my opinion.

thiryr commented on addons/website_event_booth/static/src/scss/website_event_booth.scss:
```diff
@@ -71,3 +71,13 @@ label.o_wbooth_category_unavailable {
     background-color: $o-wevent-bg-color-light;
 }
+
+.o_wevent_cta  {
```
If it's always the same, can't we put it at the root to avoid copying it everywhere? website_event will always be installed for these modules so maybe just have it there?

thiryr reviewed (commented):
Couldn't find any further issue with the styling, after these comments I think we're ok.

Thanks for the quick changes :)

xlu-odoo commented:
@thiryr / @stefanorigano - If you can take a last look here guys, thanks 💪

thiryr commented on addons/website_event/views/event_templates_list.xml:
```diff
@@ -147,11 +147,23 @@
     <t t-else="" t-set="opt_event_size" t-value="opt_index_sidebar and 'col-12' or 'col-xl-10 offset-xl-1'"/>
     <!-- No events -->
-    <t t-if="not event_ids">
+    <t t-if="not event_ids"> 
```
```suggestion
    <t t-if="not event_ids">
```
useless diff

thiryr reviewed (approved):
Ok for me, thanks for the work :)

stefanorigano commented on addons/website_event/static/src/scss/event_templates_common.scss:
```diff
@@ -193,3 +193,18 @@ $o-wevent-color: mix(map-get($theme-colors, 'primary'), $body-color, $o-wevent-c
     }
 }
+
+.o_wevent_cta  {
```
Why not using `.btn-primary` directly ? 🤔
Beside saving a couple of css lines, using the right class will ensure that the button is readable regardless by theme customizations

xlu-odoo replied:
As discussed, we need to create a new design for these components that are rendered inside the page but that are not part of the content.

stefanorigano replied:
Yes, ok,  thanks

stefanorigano commented on addons/website_event/views/event_svg.xml:
```diff
@@ -0,0 +1,49 @@
+<odoo>
+    <template id="website_event.event_svg">
+        <svg width="211" height="179" viewBox="0 0 211 179" fill="none" xmlns="http://www.w3.org/2000/svg">
+            <g fill="var(--o-color-1, #000)" >
```
Unnecessary empty space in almost all closing tags
```suggestion
            <g fill="var(--o-color-1, #000)">
```

stefanorigano commented on addons/website_event_booth/views/event_booth_templates.xml:
```diff
@@ -22,22 +22,38 @@
-                <div t-if="event.is_finished" class="container d-flex flex-column flex-grow-1">
-                    <span class="alert alert-info">This event is finished. It's no longer possible to book a booth.</span>
+                <div t-if="event.is_finished">
+                    <div class="col-12">
```
You have a `col-x` item that's not a `.row` direct descendant.
Either `col-12` is unnecessary or a `.row` is missing. Could you have a look?

stefanorigano commented on addons/website_event_booth/views/event_booth_templates.xml:
```diff
@@ -22,22 +22,38 @@
-                    <span class="alert alert-info">This event is finished. It's no longer possible to book a booth.</span>
+                <div t-if="event.is_finished">
+                    <div class="col-12">
+                        <div class="d-flex flex-column align-items-center justify-content-center text-center">
```
These flex classes shouldn't be necessary... 
The first `<div>` will take an entire row, the following `<h2>` will take another one (both are `display: block` by default).
 could you check?

stefanorigano commented on addons/website_event_booth/views/event_booth_templates.xml:
```diff
@@ -22,22 +22,38 @@
+                            <h2 class="mb-2">Registration Not Open</h2>
+                            <p>This event is not open to exhibitors registration at this time</p>
+                        </div>
+                        <div class="o_not_editable d-flex flex-column align-items-center justify-content-between my-3" groups="event.group_event_user">
```
All these flex classes are probably unnecessary

stefanorigano commented on addons/website_event_booth/views/event_booth_templates.xml:
```diff
@@ -22,22 +22,38 @@
-                    </span>
+                <div t-elif="not event_booths">
+                    <div class="col-12">
+                        <div class="d-flex flex-column align-items-center justify-content-center text-center">
```
Unnecessary flex classes, `text-center` seems enough

stefanorigano commented on addons/website_event_booth/views/event_booth_templates.xml:
```diff
@@ -22,22 +22,38 @@
+                                </a>
+                            </div>
+                            <div t-else="">
+                                Check our <a href="/event">list of future events</a>.
```
links should always have a `title`attribute (and `aria-label`, especially for the frontend)

stefanorigano commented on addons/website_event_exhibitor/views/event_exhibitor_templates_list.xml:
```diff
@@ -121,13 +121,18 @@
-            <div class="h2 mb-3">No exhibitor found.</div>
-            <div t-if="search_key" class="alert alert-info text-center">
-                <p class="m-0">We did not find any exhibitor matching your <strong t-out="search_key"/> search.</p>
+            <div class="d-flex flex-column align-items-center justify-content-center text-center">
```
Here again, flex classes are probably unnecessary ... could you check all the occurrences?

stefanorigano commented on addons/website_event_track/views/event_track_templates_list.xml:
```diff
@@ -117,13 +117,18 @@
-                    <p class="m-0">Schedule some tracks to get started</p>
+            <div class="o_not_editable d-flex flex-column align-items-center justify-content-between my-3" groups="event.group_event_user">
+                <a class="o_wevent_cta btn" target="_blank" t-att-href="'/web?#action=website_event_track.action_event_track_from_event&amp;active_id=%s' % event.id">
+                    <span class="fa fa-plus me-1"/>
```
Convention: wherever it makes sense, code **inline content** like it will be rendered, so:
```xml
<a>
    <span class="icon"> My awesome text
</a>
```

stefanorigano reviewed (changes requested):
Hello @xlu-odoo , sorry for the long delay.

xlu-odoo commented:
@stefanorigano

stefanorigano commented:
@xlu-odoo , your branch has conflicts, could you rebase? thanks

stefanorigano commented on addons/website_event_booth/views/event_booth_templates.xml:
```diff
@@ -22,23 +22,42 @@
-                    </span>
-                </div>
+                <t t-if="event.is_finished">
+                    <div class="container">
```
FYI, no need to change, but next time you can apply if conditions in the html tag directly, eg

```xml
<div t-if="event.is_finished" class="container">
```

stefanorigano reviewed (approved):
Thanks @xlu-odoo , LGTM 👍 

@odoo/rd-sm-customer , ready for tech review

stefanorigano commented:
@odoo/rd-sm-customer , reminder, thanks :)

tde-banana-odoo commented:
@stefanorigano as we are in the freeze, we will probably check this one right after it.

thiryr commented on addons/website_event_booth/views/event_booth_templates.xml:
```diff
@@ -22,23 +22,42 @@
+                                </div>
+                                <h2>Registration Not Open</h2>
+                                <p>This event is not open to exhibitors registration at this time</p>
+                                <div class="o_not_editable my-3" groups="event.group_event_user">
+                                    <div t-if="request.env.user.has_group('event.group_event_manager')">
```
One of these group checks should not be there. I would use `groups="event.group_event_manager"` as only managers are allowed to create booths. I couldn't find any justification for any of the few places that use `has_group()` instead of `groups=`, and this seems to work and be used elsewhere in the template so it should be fine to go with either. Up to you.

(I'm partial to just using `groups=` everywhere though)

thiryr reviewed (commented):
Technically all still looks fine to me, there isn't much change on that front. Just a couple comments.

Leaving a couple more functional suggestions.

xlu-odoo commented on addons/website_event/views/event_templates_list.xml:
```diff
@@ -148,10 +148,17 @@
+                <t t-call="website_event.event_svg"/>
+            </div>
+            <h3>No events scheduled yet</h3>
+            <p t-if="searches and searches['search']">We couldn't find any event matching your <strong t-out="searches and searches['search']"/> search.</p>
```
👋  Thanks! Updated

Tried something based on your suggestion, could you check if it is seems correct to you?

thiryr reviewed (approved):
Optional comment otherwise I think we're good.

Thanks for the changes!

tde-banana-odoo reviewed (commented):
Thanks for the PR :) globally LGTM, added some linting comments, approving @stefanorigano proposals to nest qweb primitives when possible, we gain an indent level :) 

Cheers !

xlu-odoo commented:
> Thanks for the PR :) globally LGTM, added some linting comments, approving @stefanorigano proposals to nest qweb primitives when possible, we gain an indent level :)
> 
> Cheers !

Thanks for the review @tde-banana-odoo ! Updated with all your recommendations, and a few others place that I caught (nesting).

tde-banana-odoo commented:
@robodoo r+

Thanks for the followup :) !
